vault/logical_system.go

```diff
@@ -1418,13 +1437,16 @@ func (b *SystemBackend) handleTuneWriteCommon(ctx context.Context, path string,
 	}
 
 	if rawVal, ok := data.GetOk("audit_non_hmac_request_keys"); ok {
-		auditNonHMACRequestKeys := rawVal.([]string)
+		newVal := rawVal.([]string)
+		auditNonHMACRequestKeys, err := augmentTuneWriteParams(newVal)
```

Comment:
Why do we need this?  I thought this part was already working, since the param is declared as a TypeCommaStringSlice?

Reply:
Yes, in the schema, the fields are defined as TypeCommaStringSlice. However, it seems that there is a bug somewhere in the mapstructure where the input is decoded. So, here is the code path: 
data.GetOk.-> data.GetOkError -> data.getPrimitive -> parseutil.ParseCommaStringSlice.

In there, mapstructure.DecoderConfig is instantiated having DecodeHook set to  mapstructure.StringToSliceHookFunc(","). There is a check in StringToSliceHookFunc where the bug show its symptom:
   if f != reflect.String || t != reflect.Slice {
       return data, nil
  }


"f" is defined the type for the input data, and "t" is defined the type for the output data. "f" should be string and "t" should be slice. 

As it seems, the "t" should be set in the DecoderConfig and should not be changed throughout the execution of the decode. However, it seems that it changes to something other than a slice, like a string. And this is where the comma separated strings are not converted properly (because the code returns before it reaches "strings.Split(raw, sep)") . So, the execution goes as follows:

From parseutil.ParseCommaStringSlice -> decoder.Decode -> d.decode : from this point on, some sort of recursive calls go through each element of the input where the output type changes from slice to string. And, if this change does not happen, the execution just loops for ever. It is hard to document all the steps here, let's have a chat when you have time to go through the sequence of function executions. 

I have not found a fix yet in a way that nothing else breaks. :/
So, long story short, the existing old code seems to be not working if the input is some sort of a slice, and that is why if we pass in "foo,bar" in the curl command, it works, but passing ["foo,bar"] it does not. For the cli, it always converts the input to slice and sends the slice to the endpoint.

Here are some examples of the results. I reverted my change for "-audit-non-hmac-request-keys" but kept it for "-allowed-response-headers" to show the difference in the output:
Command: 
vault auth tune   -audit-non-hmac-request-keys="value5,value6" -allowed-response-headers="authorization,www-authenticate" userpass

Result in Json format:
  "userpass/": {
    "uuid": "b01d4748-8490-e357-fb27-a8d4fad81508",
    "type": "userpass",
    "description": "",
    "accessor": "auth_userpass_44f766ab",
    "config": {
      "default_lease_ttl": 0,
      "max_lease_ttl": 0,
      "force_no_cache": false,
      "audit_non_hmac_request_keys": [
        "value5,value6"
      ],
      "passthrough_request_headers": [
        "Authorization"
      ],
      "allowed_response_headers": [
        "authorization",
        "www-authenticate"
      ],
      "token_type": "default-service"
    },
    "options": null,
    "local": false,
    "seal_wrap": false,
    "external_entropy_access": false
  }

Please let me know if you have any suggestions or solutions other than the approach I took to rerun ParseCommaStringSlice for every element in a slice.

Reply:
I may have been influenced my intuition that this was how Vault behaves when, previously, I suggested that you should only be expanding comma-separated values when a single string was provided.

I worry that by changing this existing behaviour we may break someone's usage who depends on this somehow.  I also worry that if we start treating this TypeCommaStringSlice argument this way, other endpoints that make use of TypeCommaStringSlice which don't do the extra expansion will start being viewed as bugs by some people, and changing those would also introduce backwards compatibility issues.

My hunch is that we don't want to change this.  It may be a good topic for the pod sync though, feel free to add it to the agenda doc.

Reply:
I see, sounds like a good topic for the pod sync. I will remove this part of the change for now then. Do we have some sort of a channel to understand how customers use these options? I have found some tests in our code where "Purpose" is used in KMS stanza for example, but for that we only have an example of an array like this ["foo", "bar"] which still works with this code. 
I will investigate other use cases of TypeCommaStringSlice and maybe even send a general question to a channel to see how people use that.